### PR TITLE
Line node vectorizing

### PR DESCRIPTION
**Pull request by Durman**
## Addressed problem description

#2649 #2758

## Preflight checklist

- [x] Code changes complete.
- [x] Code documentation complete.
- [x] Documentation for users complete.
- [x] Manual testing done. 
- [x] Ready for merge.

### CODE REVIEW

**Comment by Durman** [comment]
@portnov Have I done all correctly? Why the test has failed?

**Comment by portnov** [comment]
@Durman unfortunately Travis cuts the output for some reason... Try to run the tests locally. For linux it's `run_tests.sh`, for windows it's `run_tests_win.bat`. When I run
```
$ BLENDER=~/soft/blender-2.82-c2a37867f21b-linux-glibc217-x86_64/blender ./run_tests.sh 
```
at your branch, I receive

```
...FAIL: test_import_examples (tree_import_tests.ExamplesImportTest) (file='Tensile_Cover.json')
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/portnov/.config/blender/2.82/scripts/addons_contrib/sverchok/tests/tree_import_tests.py", line 104, in test_import_examples
    self.fail(error_format.format(node.name, node.bl_idname))
AssertionError: This example contains deprecated node `Line.002' (SvLineNodeMK3). Please upgrade the example file.

----------------------------------------------------------------------


```

**Comment by portnov** [comment]
So I believe you have to open that example, update the old node to new one, re-export the example and push it again.

**Comment by Durman** [comment]
@portnov I have updated json according your instruction and get the conflict. Can you give me a hint what next step should I do?))

**Comment by portnov** [comment]
I wonder how we've got this conflict. But anyway, just resolve it in your favor (drop older version) — with text editor & git on your side or with Github editor (Resolve Conflicts button)...

**Comment by Durman** [comment]
Well probably it because i had not rebuilt my local repository, now i did. Nevertheless test error again. I have run all test on my computer and here result. But it looks like it is not related with my push request. @portnov  Can you check the branch on your computer again?

<details> 
  <summary>error</summary>  

```
slice_lite.py
2020-01-12 21:36:18,175 [INFO] docs_tests: Category modifier_change: Tolerating unexistance of the documentation for the
 following nodes for now:
extrude_multi_alt.py
flip_normals.py
limited_dissolve.py
limited_dissolve_mk2.py
mesh_beautify.py
mesh_separate_mk2.py
triangulate_heavy.py
2020-01-12 21:36:18,180 [INFO] docs_tests: Category modifier_make: Tolerating unexistance of the documentation for the f
ollowing nodes for now:
csg_booleanMK2.py
join_tris.py
2020-01-12 21:36:18,183 [INFO] docs_tests: Category network: Tolerating unexistance of the documentation for the followi
ng nodes for now:
udp_client.py
2020-01-12 21:36:18,187 [INFO] docs_tests: Category number: Tolerating unexistance of the documentation for the followin
g nodes for now:
easing.py
2020-01-12 21:36:18,191 [INFO] docs_tests: Category object_nodes: Tolerating unexistance of the documentation for the fo
llowing nodes for now:
closest_point_on_mesh2.py
color_uv_texture.py
custom_mesh_normals.py
filter_blenddata.py
object_raycast2.py
points_from_uv_to_mesh.py
scene_raycast2.py
sculpt_mask.py
set_blenddata2.py
sort_blenddata.py
vertex_colors_mk3.py
2020-01-12 21:36:18,201 [INFO] docs_tests: Category scene: Tolerating unexistance of the documentation for the following
 nodes for now:
blenddata_to_svdata2.py
BMOperatorsMK2.py
cache.py
create_bvh_tree.py
curve_in.py
instancer.py
node_remote.py
objects_in_lite.py
obj_edit.py
particles.py
particles_MK2.py
uv_texture.py
2020-01-12 21:36:18,222 [INFO] docs_tests: Category script: Tolerating unexistance of the documentation for the followin
g nodes for now:
multi_exec.py
sn_functor_b.py
topology_simple.py
2020-01-12 21:36:18,231 [INFO] docs_tests: Category text: Tolerating unexistance of the documentation for the following
nodes for now:
text_out_mk2.py
2020-01-12 21:36:18,239 [INFO] docs_tests: Category vector: Tolerating unexistance of the documentation for the followin
g nodes for now:
color_in_mk1.py
color_out_mk1.py
formula_color.py
formula_deform_MK2.py
interpolation_mk2.py
2020-01-12 21:36:18,251 [INFO] docs_tests: Category viz: Tolerating unexistance of the documentation for the following n
odes for now:
lamp_out.py
vd_matrix.py
viewer_curves.py
viewer_idx28.py
viewer_polyline.py
viewer_skin.py
viewer_texture.py
viewer_texture_lite.py
viewer_typography.py
sv: remove app.handlers
{}
Reloading: <module 'bl_operators' from 'C:\\Program Files\\Blender Foundation\\Blender 2.81\\2.81\\scripts\\startup\\bl_
operators\\__init__.py'>
Reloading: <module 'bl_ui' from 'C:\\Program Files\\Blender Foundation\\Blender 2.81\\2.81\\scripts\\startup\\bl_ui\\__i
nit__.py'>
Reloading: <module 'keyingsets_builtins' from 'C:\\Program Files\\Blender Foundation\\Blender 2.81\\2.81\\scripts\\start
up\\keyingsets_builtins.py'>
Reloading: <module 'nodeitems_builtins' from 'C:\\Program Files\\Blender Foundation\\Blender 2.81\\2.81\\scripts\\startu
p\\nodeitems_builtins.py'>
Warning: addon-module add_mesh_extra_objects found module but without '__addon_enabled__' field, possible name collision
 from file: 'C:\\Program Files\\Blender Foundation\\Blender 2.81\\2.81\\scripts\\addons\\add_mesh_extra_objects\\__init_
_.py'
sv: import settings
sv: import all modules
sv: append app.handlers
sv: enable internal debug logging.
2020-01-12 21:36:20,006 [INFO] sverchok.utils.logging: log level, 20
Can't obtain buffer
 ______ _    _ _______  ______ _______ _     _  _____  _     _
/______  \  /  |______ |_____/ |       |_____| |     | |____/
______/   \/   |______ |    \_ |_____  |     | |_____| |    \_
initialized.

sv: version: 0.6.0.0, branch line_node_vectorizing, commit b7895a6d
sv: 345 nodes.
Reload complete, press update
gc.collect() -> 14330
2020-01-12 21:36:20,168 [INFO] init_tests: Disabling Sverchok
sv: remove app.handlers
{}
2020-01-12 21:36:20,177 [INFO] init_tests: Enabling Sverchok
sv: append app.handlers
sv: enable internal debug logging.
2020-01-12 21:36:20,182 [INFO] sverchok.utils.logging: log level, 20
Can't obtain buffer
 ______ _    _ _______  ______ _______ _     _  _____  _     _
/______  \  /  |______ |_____/ |       |_____| |     | |____/
______/   \/   |______ |    \_ |_____  |     | |_____| |    \_
initialized.

sv: version: 0.6.0.0, branch line_node_vectorizing, commit b7895a6d
sv: 345 nodes.
Reload complete, press update
get_and_set_gl_scale_info called from sv_post_load
getting gl scale params
set render_scale to: 1.5
set render_location_xy_multiplier to: 1.5
2020-01-12 21:36:20,388 [INFO] profile_mk3_tests: Checking profile example: arc_fillet.txt
2020-01-12 21:36:20,399 [INFO] profile_mk3_tests: Checking profile example: barrel.txt
2020-01-12 21:36:20,410 [INFO] profile_mk3_tests: Checking profile example: bevel_line.txt
2020-01-12 21:36:20,420 [INFO] profile_mk3_tests: Checking profile example: complex_curve.txt
2020-01-12 21:36:20,433 [INFO] profile_mk3_tests: Checking profile example: complex_rectangle.txt
2020-01-12 21:36:20,446 [INFO] profile_mk3_tests: Checking profile example: i-beam.txt
2020-01-12 21:36:20,457 [INFO] profile_mk3_tests: Checking profile example: pawn.txt
2020-01-12 21:36:20,471 [INFO] profile_mk3_tests: Checking profile example: quadratic_bubble.txt
2020-01-12 21:36:20,481 [INFO] profile_mk3_tests: Checking profile example: quadratic_fillet.txt
2020-01-12 21:36:20,491 [INFO] profile_mk3_tests: Checking profile example: quadratic_goggles.txt
2020-01-12 21:36:20,503 [INFO] profile_mk3_tests: Checking profile example: quadratic_sample.txt
2020-01-12 21:36:20,515 [INFO] profile_mk3_tests: Checking profile example: rectangle.txt
2020-01-12 21:36:20,524 [INFO] profile_mk3_tests: Checking profile example: rounded_line.txt
2020-01-12 21:36:20,535 [INFO] profile_mk3_tests: Checking profile example: simple_curve.txt
2020-01-12 21:36:20,545 [INFO] profile_mk3_tests: Checking profile example: smooth_arrow.txt
2020-01-12 21:36:20,556 [INFO] profile_mk3_tests: Checking profile example: smooth_curve.txt
2020-01-12 21:36:20,568 [INFO] profile_mk3_tests: Checking profile example: square.txt
2020-01-12 21:36:20,578 [INFO] profile_mk3_tests: Checking profile example: square3.txt
2020-01-12 21:36:20,588 [INFO] profile_mk3_tests: Checking profile example: symmetric_square.txt
2020-01-12 21:36:20,598 [INFO] profile_mk3_tests: Checking profile example: zeffi_1.txt
2020-01-12 21:36:20,613 [INFO] profile_mk3_tests: Checking profile example: zeffi_2.txt
2020-01-12 21:36:20,719 [INFO] sverchok.utils.sv_IO_panel_tools: ############0.079
2020-01-12 21:36:20,743 [INFO] sverchok.utils.sv_IO_panel_tools: ############0.079
2020-01-12 21:36:20,768 [INFO] sverchok.utils.sv_IO_panel_tools: ############0.079
1
0 ('SvStringsSocket', 'arc', 1.0, 2)
1 ('SvStringsSocket', 'separation', 3.0, 2)
2 ('SvStringsSocket', 'num_points', 300, 2)
3 ('SvStringsSocket', 'scale', 0.13, 2)
[]
2020-01-12 21:36:21,043 [INFO] sverchok.utils.sv_IO_panel_tools: ############0.079
2020-01-12 21:36:21,075 [INFO] sverchok.utils.sv_IO_panel_tools: ############0.079
2020-01-12 21:36:21,081 [INFO] sverchok.utils.sv_IO_panel_tools: ############0.079
sockets wiped, skipped update
sockets wiped, skipped update
2020-01-12 21:36:21,155 [INFO] sverchok.utils.sv_IO_panel_tools: ############0.079
2020-01-12 21:36:21,248 [INFO] sverchok.utils.sv_IO_panel_tools: ############0.079
0 ('SvStringsSocket', 'arc', 1.0, 2)
1 ('SvStringsSocket', 'separation', 3.0, 2)
2 ('SvStringsSocket', 'num_points', 300, 2)
3 ('SvStringsSocket', 'scale', 0.13, 2)
2020-01-12 21:36:22,159 [INFO] sverchok.utils.testing: Test cases result:
test_avl_tree_find_nearest_left (avl_tree_tests.AVLTreeTest) ... ok
test_avl_tree_search (avl_tree_tests.AVLTreeTest) ... ok
test_eval (cubic_spline_tests.CubicSplineTests) ... ok
test_tangent (cubic_spline_tests.CubicSplineTests) ... ok
test_calc_mask_1 (data_structure_tests.CalcMaskTests) ... ok
test_calc_mask_2 (data_structure_tests.CalcMaskTests) ... ok
test_calc_mask_3 (data_structure_tests.CalcMaskTests) ... ok
test_calc_mask_4 (data_structure_tests.CalcMaskTests) ... ok
test_calc_mask_5 (data_structure_tests.CalcMaskTests) ... ok
test_calc_mask_6 (data_structure_tests.CalcMaskTests) ... ok
test_calc_mask_7 (data_structure_tests.CalcMaskTests) ... ok
test_describe_data_shape (data_structure_tests.DataStructureTests) ... ok
test_ensure_nesting_level_1 (data_structure_tests.DataStructureTests) ... ok
test_ensure_nesting_level_2 (data_structure_tests.DataStructureTests) ... ok
test_full_list_1 (data_structure_tests.DataStructureTests) ... ok
test_full_list_2 (data_structure_tests.DataStructureTests) ... ok
test_full_list_deep_copy (data_structure_tests.DataStructureTests) ... ok
test_get_data_nesting_level_1 (data_structure_tests.DataStructureTests) ... ok
test_get_data_nesting_level_2 (data_structure_tests.DataStructureTests) ... ok
test_match_long_cycle (data_structure_tests.DataStructureTests) ... ok
test_match_long_repeat (data_structure_tests.DataStructureTests) ... ok
test_rotate_list_1 (data_structure_tests.DataStructureTests) ... ok
test_rotate_list_2 (data_structure_tests.DataStructureTests) ... ok
test_transpose_list (data_structure_tests.DataStructureTests) ... ok
test_diameter_1 (diameter_tests.DiameterTests) ... ok
test_diameter_2 (diameter_tests.DiameterTests) ... ok
test_diameter_3 (diameter_tests.DiameterTests) ... ok
test_all_node_docs_in_trees (docs_tests.DocumentationTests) ... ok
test_node_docs_existance (docs_tests.DocumentationTests) ... ok
test_node_docs_refs_from_trees (docs_tests.DocumentationTests) ... ok
test_unused_docs (docs_tests.DocumentationTests) ... ok
test_crop_ring_by_squares (geom_2d_tests.CropMesh2DTest) ... ok
test_crop_voronoi_diagram (geom_2d_tests.CropMesh2DTest) ... ok
test_dissolve_ngon_faces_inside_grid (geom_2d_tests.DissolveFaces2DTest) ... ok
test_holes_in_holes (geom_2d_tests.EdgesToFacesTest) ... ok
test_random_boxes_intersection (geom_2d_tests.EdgesToFacesTest) ... ok
test_random_edges_intersection (geom_2d_tests.EdgesToFacesTest) ... ok
test_intersect_edges_in_one_point (geom_2d_tests.IntersectionEdgesTest) ... ok
test_intersect_overlapping_edges (geom_2d_tests.IntersectionEdgesTest) ... ok
test_intersect_overlapping_points_with_edges (geom_2d_tests.IntersectionEdgesTest) ... ok
test_intersect_random_edges (geom_2d_tests.IntersectionEdgesTest) ... ok
test_random_edges_above_xy (geom_2d_tests.IntersectionEdgesTest) ... ok
test_random_edges_turned (geom_2d_tests.IntersectionEdgesTest) ... ok
test_make_monotone_one_face (geom_2d_tests.MakeMonotoneTest) ... ok
test_merge_random_squares (geom_2d_tests.MergeMesh2DLightTest) ... ok
test_merge_rings_and_square (geom_2d_tests.MergeMesh2DLightTest) ... ok
test_merge_rings_and_square_turned (geom_2d_tests.MergeMesh2DLightTest) ... ok
test_number_overlapping_of_flour (geom_2d_tests.MergeMesh2DLightTest) ... ok
test_merge_square_ring (geom_2d_tests.MergeMesh2DTest) ... ok
test_center_quad (geom_tests.GeometryTests) ... ok
test_center_trivial (geom_tests.GeometryTests) ... ok
test_normal_quad (geom_tests.GeometryTests) ... ok
test_core_exists (init_tests.InitTests) ... ok
test_disable_enable (init_tests.InitTests) ... ok
test_intersect_edges (intersect_edges_tests.IntersectEdgesTest2) ... ok
test_check_no (linear_approximation_tests.LineTests) ... ok
test_check_yes (linear_approximation_tests.LineTests) ... ok
test_distance_to_point (linear_approximation_tests.LineTests) ... ok
test_from_two_points (linear_approximation_tests.LineTests) ... ok
test_approximate_line_1 (linear_approximation_tests.LinearApproximationTests) ... ok
test_approximate_line_2 (linear_approximation_tests.LinearApproximationTests) ... ok
test_approximate_plane (linear_approximation_tests.LinearApproximationTests) ... ok
test_check_no (linear_approximation_tests.PlaneTests) ... ok
test_check_yes (linear_approximation_tests.PlaneTests) ... ok
test_distance_to_point (linear_approximation_tests.PlaneTests) ... ok
test_distance_to_points (linear_approximation_tests.PlaneTests) ... ok
test_intersect_with_line (linear_approximation_tests.PlaneTests) ... ok
test_intersect_with_plane (linear_approximation_tests.PlaneTests) ... ok
test_nearest_to_origin (linear_approximation_tests.PlaneTests) ... ok
test_plane_from_three_points (linear_approximation_tests.PlaneTests) ... ok
test_two_vectors (linear_approximation_tests.PlaneTests) ... ok
test_eval (linear_spline_tests.LinearSplineTests) ... ok
test_tangent (linear_spline_tests.LinearSplineTests) ... ok
test_ngon_verts_1 (ngon_tests.NGonTests) ... ok
test_ngon_verts_2 (ngon_tests.NGonTests) ... ok
test_import_examples (profile_mk3_tests.ExamplesParseTests) ... ok
test_expr (profile_mk3_tests.SimpleTests) ... ok
test_identifier (profile_mk3_tests.SimpleTests) ... ok
test_negated_var (profile_mk3_tests.SimpleTests) ... ok
test_close_All (profile_mk3_tests.StatementParseTests) ... ok
test_close_path (profile_mk3_tests.StatementParseTests) ... ok
test_parse_assign (profile_mk3_tests.StatementParseTests) ... ok
test_parse_curveto (profile_mk3_tests.StatementParseTests) ... ok
test_parse_default (profile_mk3_tests.StatementParseTests) ... ok
test_parse_hor_lineto (profile_mk3_tests.StatementParseTests) ... ok
test_parse_hor_lineto_n (profile_mk3_tests.StatementParseTests) ... ok
test_parse_lineto (profile_mk3_tests.StatementParseTests) ... ok
test_parse_lineto_n (profile_mk3_tests.StatementParseTests) ... ok
test_parse_moveto (profile_mk3_tests.StatementParseTests) ... ok
test_parse_vert_lineto (profile_mk3_tests.StatementParseTests) ... ok
test_parse_vert_lineto_n (profile_mk3_tests.StatementParseTests) ... ok
test_box_export (simple_export_tests.ExportSingleSimpleNode) ... ok
test_cylinder_export (simple_export_tests.ExportSingleSimpleNode) ... ok
test_torus_export (simple_export_tests.ExportSingleSimpleNode) ... ok
test_box (simple_generator_tests.BoxNodeTest) ... ok
test_ngon (simple_generator_tests.NGonNodeTest) ... ok
test_box_import (simple_import_tests.ImportSingleSimpleNode) ... ok
test_cylinder_import (simple_import_tests.ImportSingleSimpleNode) ... ok
test_torus_import (simple_import_tests.ImportSingleSimpleNode) ... ok
test_text_viewer (simple_viewer_text_tests.TextViewerTest) ... FAIL
test_adaptive_sockets (socket_conversion_tests.SocketConversionTests) ... ok
test_vertices_to_matrices (socket_conversion_tests.SocketConversionTests) ... ok
test_mesh_expr_export (tree_export_tests.MeshExprExportTest) ... ok
test_monad_export (tree_export_tests.MonadExportTest) ... skipped 'Linking node tree with Monad node does not work corre
ctly.'
test_profile_export (tree_export_tests.ProfileExportTest) ... ok
test_script_uv_export (tree_export_tests.ScriptUvExportTest) ... ok
test_textview_expr_export (tree_export_tests.ViewerTextExportTest) ... ok
test_import_examples (tree_import_tests.ExamplesImportTest) ... skipped 'This test is intended for batch mode only'
test_mesh_expr_import (tree_import_tests.MeshExprImportTest) ... ok
test_monad_import (tree_import_tests.MonadImportTest) ... ok
test_profile_import (tree_import_tests.ProfileImportTest) ... ok
test_script_uv_import (tree_import_tests.ScriptUvImportTest) ... ok
test_all_nodes_have_icons (ui_tests.UiTests) ... ok
test_make_dep_dict (update_system_tests.UpdateSystemTests) ... ok
test_make_update_list (update_system_tests.UpdateSystemTests) ... ok

======================================================================
FAIL: test_text_viewer (simple_viewer_text_tests.TextViewerTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\User\YandexDisk\3.╨б╨▓╨╛╨╕_╨┐╤А╨╛╨╡╨║╤В╤Л\Blender_code\addons\sverchok\tests\simple_viewer_text_tests.p
y", line 29, in test_text_viewer
    self.assertEquals(text, expected_text)
AssertionError: 'node[280 chars]865476, -1.0]\n[0.5555702330196023, 0.83146961[2840 chars]    ' != 'node[280 chars]86547
5, -1.0]\n[0.5555702330196023, 0.83146961[2842 chars]    '
Diff is 3972 characters long. Set self.maxDiff to None to see it.

----------------------------------------------------------------------
Ran 115 tests in 4.108s

FAILED (failures=1, skipped=2)

Error   : EXCEPTION_ACCESS_VIOLATION
Address : 0x000000013FD2B643
Module  : C:\Program Files\Blender Foundation\Blender 2.81\blender.exe
Для продолжения нажмите любую клавишу . . .
```

</details>

**Comment by portnov** [comment]
Hmm... tests are running okay for me now...

**Comment by portnov** [comment]
Now Travis shows the output, but it is not expected — there is still deprecated node...

**Comment by portnov** [comment]
Uhh, sorry, it appears I've ran tests on wrong repo copy... :)

**Comment by portnov** [comment]
Aha, now my output is the same as Travis's...
```
======================================================================
FAIL: test_import_examples (tree_import_tests.ExamplesImportTest) (file='Color_Lights.json')
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/portnov/.config/blender/2.82/scripts/addons_contrib/sverchok/tests/tree_import_tests.py", line 104, in test_import_examples
    self.fail(error_format.format(node.name, node.bl_idname))
AssertionError: This example contains deprecated node `Line MK2' (SvLineNodeMK3). Please upgrade the example file.

======================================================================
FAIL: test_import_examples (tree_import_tests.ExamplesImportTest) (file='Tensile_Cover.json')
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/portnov/.config/blender/2.82/scripts/addons_contrib/sverchok/tests/tree_import_tests.py", line 100, in test_import_examples
    import_tree(new_tree, path)
  File "/home/portnov/SSD/soft/blender-2.82-c2a37867f21b-linux-glibc217-x86_64/2.82/python/lib/python3.7/contextlib.py", line 119, in __exit__
    next(self.gen)
  File "/home/portnov/.config/blender/2.82/scripts/addons_contrib/sverchok/utils/testing.py", line 491, in assert_logs_no_errors
    self.assertFalse(has_errors, "There were some errors logged")
AssertionError: True is not false : There were some errors logged

----------------------------------------------------------------------
Ran 115 tests in 5.291s

FAILED (failures=2, skipped=1)
```

**Comment by portnov** [comment]
:+1:

**Comment by vicdoval** [comment]
image
The AB method with Normalize activated was working as a OP mode (Origin, point in line), although this can be achieved with the help of other nodes I don't know if it is possible to make an exact replacement with the replacement operator  

I see your point in splitting the node, but finally they are both straight segment's lists, none of the nodes produce a true line and both produce  line-like geometry, why one should be called segment?
I was thinking about how the icons should be for each one, but the output can be identical... 

This node-tree is possible now:
image
[line_test1.zip](https://github.com/nortikin/sverchok/files/4063791/line_test1.zip)

So I propose you to join them again into one single node with all the possibilities inside

Anothe minor issues:

The Num Verts of the AB mode do not get updated to  the Segment Count
When replacing the previous node from the OD mode this error pops in the console and no mode is selected
` line 93, in get_corner_points
    if dir_mode == DIRECTION.od else np.array(directions[dir_mode]))
KeyError: ''`

**Comment by vicdoval** [comment]
Also this is possible with the actual implementation (size determined and controlled step)
image

**Comment by Durman** [comment]
> The AB method with Normalize activated was working as a OP mode (Origin, point in line)

The same logic can be repeated with `OD` mode.


> Also this is possible with the actual implementation (size determined and controlled step)

If steps will be prepared for step mode the same can be done.


Is it really worth to implement all this logic inside the node?

> none of the nodes produce a true line and both produce line-like geometry, why one should be called segment?

What do you mean? That all nodes which produce the same data should be combined in one node? If so I can't be agree less with such approach. Have a look to matrix category. We have a lot of nodes there which output is just matrixes. But all of them has different algorithm of creating them. Just the same with line and segment nodes. 

Probably you can suggest more correct names for this nodes.

> The Num Verts of the AB mode do not get updated to the Segment Count

What is your proposal? I can delete the `replacement_nodes` attribute so users can do replacement just manually.

**Comment by vicdoval** [comment]
The node was working like that and probably some of the features are maybe useless or are achievable with more nodes, I am just trying not to loose anything on the upgrade. 
Every time you add a node to the tree it gets slower and you have to find it, connect it and so, that is why I got it to coding nodes... so if to replace a functionality you need 1 or 2 more nodes  i would say it worth implementing it.
As i see it there where three modes in this node Ortho Line, 2pt Line and OD Line, but they can share multiple options so it was cool to have it one node.
There was this PR #2318 that was reducing the node to AB and OD, which I believe is a good idea.
AB  and OD is not too much for a node I think but if you want to split it i would go for something like 2pt Line and Origin-Direction Line (or OD Line) and even XYZ Line
All this it is just one opinion maybe someone else any a thoughts about this? @zeffii @portnov @DolphinDream @nortikin

**Comment by zeffii** [comment]
@vicdoval  do you feel there's a regression here, and it impedes the most convenient approach..

i don't really have my ear to the railtrack on this issue.

**Comment by portnov** [comment]
If you look at Grasshopper, there is a separate node for each separate function, and even for the same function performed on different types of objects. So, this approach "has a right to live". But it seems that Sverchok's approach differs in most cases. Sverchok tends to have more universal nodes. But, not too universal — that would make them too complicated to maintain and to learn.

**Comment by Durman** [comment]
> As i see it there where three modes in this node Ortho Line, 2pt Line and OD Line.

There was no three modes, only two. `X`, `Y`, `Z` modes are corner cases of `OD` mode with origin in center of coordinate system and `XYZ` directions. 

The reason why I have splitted the node is that their does not share the same input properties. The sense of `X`, `Y`, `Z` and `OD` modes is to create line starting from origin with custom direction. And there are some modes for determining its length. 

The sense of `AB` mode is to create line between two points. This is different approach. There is no sense to have options for determining its length because it is already determined by two points.

**Comment by Durman** [comment]
Voting for name of the node:

1. Edge
2. Segment
3. 2pt Line
4. Origin-Direction Line
5. OD Line
6. XYZ Line

**Comment by Durman** [comment]
I hope that now I have covered all hidden features of previous node. 😄

File: old_nodes/line_mk3.py [hunk 1]
```diff
@@ -90,46 +90,53 @@ def update_vect_socket(self, context):
 
         updateNode(self, context)
 
-    direction : EnumProperty(
+    @property
+    def replacement_nodes(self):
+        if self.direction == 'AB':
```
Comment (vicdoval) [inline old_nodes/line_mk3.py]: if self.direction == 'AB' and not self.normalize

**Pull request merged**